Integrate logcompaction table service on metadata table and provides various bugfixes to metadata table

suryaprasanna opened the pull request:
Following changes are also included as part of this commit:

1. Change cleaner policy to KEEP_LATEST_FILE_VERSIONS for metadata table
2. Added logs statements
3. Fix rollback of inflight deltacommits on metadata table
4. Add log statements to LogCompaction planner
5. Integrate logcompaction table service on metadata table
6. Include checks to block Logcompaction in presence of compaction with greater timestamp
7. Use instant range option for major and minor compaction on metadata table

## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contributing.html before opening a pull request.*

## What is the purpose of the pull request

*(For example: This pull request adds quick-start document.)*

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

suryaprasanna committed:
Integrate logcompaction table service on metadata table and provides various bugfixes to metadata table

Following changes are also included as part of this commit:

1. Change cleaner policy to KEEP_LATEST_FILE_VERSIONS for metadata table
2. Added logs statements
3. Fix rollback of inflight deltacommits on metadata table
4. Add log statements to LogCompaction planner
5. Integrate logcompaction table service on metadata table
6. Include checks to block Logcompaction in presence of compaction with greater timestamp
7. Use instant range option for major and minor compaction on metadata table

Zouxxyy committed:
[MINOR] Reformat HoodieAvroParquetReader (#8888)

isnker committed:
[MINOR] Use spaces instead of tabs (#8895)

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1028,9 +1044,24 @@ private void compactIfNecessary(BaseHoodieWriteClient writeClient, String latest
-    if (!metadataMetaClient.getActiveTimeline().filterCompletedInstants().containsInstant(compactionInstantTime)
-        && writeClient.scheduleCompactionAtInstant(compactionInstantTime, Option.empty())) {
+    if (metadataMetaClient.getActiveTimeline().filterCompletedInstants().containsInstant(compactionInstantTime)) {
+      LOG.info(String.format("Compaction with same %s time is already present in the timeline.", compactionInstantTime));
```
We keep bloating this single class for a long time. We should think about adding a new class called TableServiceMetadateWriter or something and move all table service related code to that. 
I don't want to drag this patch too. but lets see if we can get something in after the major patches are landed.

suryaprasanna replied:
Sure. Created a ticket to track. https://issues.apache.org/jira/browse/HUDI-6335

codope replied:
+1
Created https://issues.apache.org/jira/browse/HUDI-6338 to track

suryaprasanna replied:
Already create a ticket you can close either of them as duplicate.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1028,9 +1044,24 @@ private void compactIfNecessary(BaseHoodieWriteClient writeClient, String latest
+      LOG.info("Compaction is scheduled for timestamp " + compactionInstantTime);
       writeClient.compact(compactionInstantTime);
+    } else if (metadataWriteConfig.isLogCompactionEnabled()) {
+      // Schedule and execute log compaction with suffixes based on the same instant time. This ensures that any future
```
So, we can do log compaction only if regular compaction does not get scheduled? I assume this is intentional ?

suryaprasanna replied:
Yes.

codope replied:
Yeah it makes sense too at least for the MDT.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1066,17 +1097,49 @@ private boolean validateTimelineBeforeSchedulingCompaction(Option<String> inFlig
-    if (!pendingInstants.isEmpty()) {
-      checkNumDeltaCommits(metadataMetaClient, dataWriteConfig.getMetadataConfig().getMaxNumDeltacommitsWhenPending());
+    // There should not be any incomplete instants on MDT
+    HoodieActiveTimeline metadataTimeline = metadataMetaClient.reloadActiveTimeline();
```
we would have reloaded the timeline just before this call at the beginning of performTableServices. 
Do we still need another reload here? 
Lets be judicious about metaclient reloads.

suryaprasanna replied:
Removed the call.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1066,17 +1097,49 @@ private boolean validateTimelineBeforeSchedulingCompaction(Option<String> inFlig
+    // There should not be any incomplete instants on dataset
+    HoodieActiveTimeline datasetTimeline = dataMetaClient.reloadActiveTimeline();
+    List<HoodieInstant> pendingInstantsOnDataset = datasetTimeline.filterInflightsAndRequested().getInstantsAsStream()
+        .filter(i -> !inFlightInstantTimestamp.isPresent() || !i.getTimestamp().equals(inFlightInstantTimestamp.get()))
```
guess you missed to resolve it correctly. 
here is what we have in master 
```
List<HoodieInstant> pendingInstants = dataMetaClient.reloadActiveTimeline().filterInflightsAndRequested()
        .findInstantsBeforeOrEquals(latestDeltaCommitTimeInMetadataTable).getInstants();
```

we find the latest delta commit in MDT. 
and then find any inflights in DT before this latest delta commit in MDT. 
we have added lot of java docs on why this is very critical. Please pay attention to these details. 
looks like we are only checking for any inflights in general. 
Or was it an intentional fix?

suryaprasanna replied:
My bad, corrected it now.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -1066,17 +1097,49 @@ private boolean validateTimelineBeforeSchedulingCompaction(Option<String> inFlig
+      metrics.ifPresent(m -> m.incrementMetric(HoodieMetadataMetrics.SKIP_TABLE_SERVICES, 1));
+      return false;
+    }
+    if (HoodieTimeline.compareTimestamps(lastCompletedInstant.get().getTimestamp(),
```
may I know why we need this specifically. we should be able to allow out of order commits. All we need to ensure is, the new compaction in MDT should have an instant time where in all commits prior to that is synced to MDT.

suryaprasanna replied:
My mistake. It was a miss from my side, OSS code fixes this correctly, reverting my changes.

suryaprasanna commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -348,7 +347,10 @@ private void initializeMetadataTable(Option<String> inFlightInstantTimestamp) {
       }
     } catch (Exception e) {
-      throw new HoodieException("Failed to instantiate Metadata table ", e);
+      LOG.error("Exception in running table services on metadata table", e);
```
Waiting for unit test build to complete. After that, I will revert this change.

suryaprasanna replied:
Reverted this change.

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/ScheduleCompactionActionExecutor.java:
```diff
@@ -178,6 +178,10 @@ private boolean needCompact(CompactionTriggerStrategy compactionTriggerStrategy)
     // get deltaCommitsSinceLastCompaction and lastCompactionTs
     Option<Pair<Integer, String>> latestDeltaCommitInfoOption = getLatestDeltaCommitInfo();
+    if (WriteOperationType.LOG_COMPACT.equals(operationType)) {
+      LOG.info("Need log compaction checks for log compaction are performed by the plan generator. So, always allow from here.");
```
Is this not same as L189 ?

suryaprasanna replied:
Yeah, good catch. Removed this.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDTableServiceClient.java:
```diff
@@ -86,13 +93,29 @@ protected HoodieWriteMetadata<JavaRDD<WriteStatus>> compact(String compactionIns
     HoodieSparkTable<T> table = HoodieSparkTable.create(config, context);
+
+    // Check if a commit or compaction instant with a greater timestamp is on the timeline.
+    // If a instant is found then abort log compaction, since it is no longer needed.
```
can you help me understand why do we need to add this? 
why can't we trigger log compaction if there are higher commits ?

suryaprasanna replied:
While creating compaction plan, filtering filegroups which are under logcompaction is false. Due to that one file group can have first log compaction plan created and later compaction plan created. In that case, when logcompaction operation writing data using HoodieAppendHandle, it will merge the records and write to latest file slice. This will create dups in new file slice. To avoid all such cases log compaction operation is failed when there is a greater compaction with greater timestamp.
We could have made compaction give up if there are pending log compaction operations, but there will be cases where compaction has to happen at a certain frequency irrespective of log compaction. Lgocompaction is mainly for performance improvements, whereas for compaction need to look into completeness and freshness.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDTableServiceClient.java:
```diff
@@ -157,7 +180,7 @@ protected void completeLogCompaction(HoodieCommitMetadata metadata,
     WriteMarkersFactory.get(config.getMarkersType(), table, logCompactionCommitTime)
         .quietDeleteMarkerDir(context, config.getMarkersDeleteParallelism());
-    if (compactionTimer != null) {
+    if (logCompactionTimer != null) {
```
shouldn't we also fix : 184 to logCompactionTimer ?

suryaprasanna replied:
Good catch. Fixed it.

nsivabalan commented on hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java:
```diff
@@ -348,7 +347,10 @@ private void initializeMetadataTable(Option<String> inFlightInstantTimestamp) {
-      throw new HoodieException("Failed to instantiate Metadata table ", e);
+      LOG.error("Exception in running table services on metadata table", e);
+      if (config.isMetricsOn()) {
+        metrics.reportMetrics("metadata.table_services.operation", "failure", 1);
```
Is there any const variable that we can reference instead of hard coding ?

suryaprasanna replied:
This is a duplicate metrics, removing it. We are already emitting metrics in HoodieBackedTableMetadataWriter finally block.

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/common/config/HoodieMetadataConfig.java:
```diff
@@ -94,6 +94,12 @@ public final class HoodieMetadataConfig extends HoodieConfig {
       .withDocumentation("Controls how often the metadata table is compacted.");
 
+  // Log blocks threshold, after a file slice crosses this threshold log compact operation is scheduled.
+  public static final ConfigProperty<Integer> LOG_COMPACT_BLOCKS_THRESHOLD = ConfigProperty
```
So, this is applicable for all partitions in MDT? 
how are we enabling this only for RLI ?

suryaprasanna replied:
Yes, this is applicable for all partitions.

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java:
```diff
@@ -469,37 +460,6 @@ private Pair<HoodieSeekingFileReader<?>, Long> getBaseFileReader(FileSlice slice
   }
 
-  public static Set<String> getValidInstantTimestamps(HoodieTableMetaClient dataMetaClient, HoodieTableMetaClient metadataMetaClient) {
-    // Only those log files which have a corresponding completed instant on the dataset should be read
```
I am assuming we just moved this code w/o any changes as such ?

suryaprasanna replied:
Yes, added logcompaction check that's all.

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/metadata/HoodieTableMetadataUtil.java:
```diff
@@ -1325,6 +1332,39 @@ public static Set<String> getInflightAndCompletedMetadataPartitions(HoodieTableC
+    validInstantTimestamps.addAll(
+        metadataMetaClient.getActiveTimeline()
+            .filter(instant -> instant.isCompleted()
+                && (isIndexingCommit(instant.getTimestamp()) || isLogCompactionInstant(instant)))
```
|| isLogCompactionInstant(instant)

this is the only change I could infer. Is there anything else?

suryaprasanna replied:
Yes, that's the only change.

codope commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieCompactionConfig.java:
```diff
@@ -61,6 +61,11 @@ public class HoodieCompactionConfig extends HoodieConfig {
           + "execution inline for compaction");
 
+  public static final ConfigProperty<String> ENABLE_LOG_COMPACTION = ConfigProperty
+      .key("hoodie.log.compaction.enable")
```
Should we rename to `hoodie.metadata.log.compaction.enable` for clarity?

suryaprasanna replied:
I understand it is little confusing if the main table is COW and metadata table is MOR.
The idea behind this config is to keep it at a table level. So, both the main table and metadata supports it or not. So, it is added this way, let me know what your thoughts are.

codope commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java:
```diff
@@ -983,10 +983,12 @@ public void performTableServices(Option<String> inFlightInstantTimestamp) {
-
+      LOG.info("No inflight commits present in either of the timelines. "
+          + "Proceeding to run rest of the table services.");
+      compactIfNecessary(writeClient, latestDeltacommitTime);
```
Should compaction be done before cleaning? In master, we compact then commit and then clean: https://github.com/apache/hudi/blob/f60afd5376f53bc109ffa018cda49b9cd3ab56be/hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/metadata/SparkHoodieBackedTableMetadataWriter.java#L159

suryaprasanna replied:
Metadata write configs are created with following configs.
            .withCleanerPolicy(HoodieCleaningPolicy.KEEP_LATEST_FILE_VERSIONS)
            .retainFileVersions(2)
This will trigger cleaner only after a compaction. 
And inside cleanIfNecessary, you have a check to trigger clean after 3 deltacommits from compaction. This is done is such a way to catch a failures and restore them. We can revisit these later once the code stabilizes.

prashantwason committed:
[HUDI-6200] Enhancements to the MDT for improving performance of larger indexes. (#8684)

Major changes

Moved MDT compaction out of the pre-commit phase into the initialization phase within write client

MDT compaction and clean is attempted during WriteClient initialization stage before any changes to the dataset are initiated.
This ensures correct ordering of the compactions
This also creates a codepath to call compaction on the MDT without actual commits
Removed support for setting fileGroup count to the MetadataPartitionType enum

Since enums are singletons, this design prevented having multiple different datasets with different fileGroup count for MDT parttitions in the same JVM instance
When there are large number of fileGroups (assume 10K+) for a MDT partition, listing fileSlices for it may take a considerable amount of time (10K slices * M basefile version + 10K * N log files per slice). Hence, this should not be done as part of the HoodieBackedTableMetadataWriter constructor and delayed to whenever actually required (prepRecords).
When initializing multiple MDT partitions together, they are handled one at a time

Reduces memory requirements as larger indexes may need a huge amount of memory
Improves chances of indexes being built.
If the FILES partition is already initialized, it is listed to find the files thereby saving expensive entire dataset listing while
initializing addtional partitions
Initial commit into MDT to initialize a new partition is not a bulkCommit

HFiles are created directly for the first commit
Since first commits write a large amount of data, bulkCommit into a basefile is much faster than writing log files
Added SparkHoodieMetadataBulkInsertPartitioner as the bulk insert partitioner. It is optimized to partition records in a single pass.
The number of fileGroups for each MDT partition are determined as part of the initialization process.

This allows the fileGroup count to be optimized based on the number of records being added
Other changes:

Created helper method to find if MDT partition is enabled: HoodieTableConfig.isMetadataPartitionEnabled()
Created helper method to set state (enabled/disabled) of MDT partition: HoodieTableConfig.setMetadataPartitionState()
Created helper method to delete MDT: HoodieTableMetadataUtil.deleteMetadataTable(). It correctly updates the hoodie.properties file.
Created helper method to set inflight state (enabled/disabled) of MDT partition: HoodieTableMetadataUtil.setMetadataInflightPartitions
Created helper method to delete MDT partition: deleteMetadataTablePartition. It correctly updates the hoodie.properties file.

xuzifu666 committed:
[HUDI-6310] CreateHoodieTableCommand::createHiveDataSourceTable arguments refactor (#8874)

Co-authored-by: xuyu <11161569@vivo.com>

xushiyan committed:
[HUDI-5352] Fix `LocalDate` serialization in colstats (#8840)

Under spark3.3 profile, when colstats is configured for Date type column, serialization will break with error in fasterxml complaining missing java time module support.  This commit fixes the issue.

---------

Co-authored-by: Y Ethan Guo <ethan.guoyihua@gmail.com>

nsivabalan committed:
[HUDI-6342] Fixing flaky Continuous mode multi writer tests (#8910)

suryaprasanna committed:
Fix test failures

suryaprasanna committed:
Refactor

suryaprasanna committed:
Refactoring

suryaprasanna committed:
Addressed review comments

suryaprasanna committed:
Address review comments

suryaprasanna committed:
Update hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieCompactionConfig.java

Co-authored-by: Sagar Sumit <sagarsumit09@gmail.com>

suryaprasanna committed:
Update hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/compact/CompactHelpers.java

Co-authored-by: Sagar Sumit <sagarsumit09@gmail.com>

suryaprasanna committed:
Remove duplicate log stmt

stream2000 committed:
[HUDI-6182] Hive sync use state transient time to avoid losing partit… (#8745)

Now Hudi would sync 2 timestamp to HMS for each sync, one is the max commit time T1(actually the max version id), another is the max completion time T2, the later can be used to trace the instants that has smaller version id than T1 but greater completion timestamp than T2. Such instants are usually generated by multi writers.

suryaprasanna committed:
Address review comments
